### PR TITLE
Generator lists outdated exercises

### PR DESCRIPTION
**Pull request by jrr**
I took a crack at https://github.com/exercism/fsharp/issues/581 .

**Generation** (`-e`) should work as before, except that now it can't be combined with `-s`.

**Status** (`-s`) now works by itself, listing exercises of the specified status. Here's how it looks with `-s outdated`:
```
[23:38:48 INF] Listing exercises with status Outdated...
[23:38:48 INF] Updating repository to latest version...
[23:38:48 INF] Updated repository to latest version.
[23:38:49 INF] 20 exercises are outdated:
[23:38:49 INF]  allergies                  (1.1.0 -> 1.2.0)
[23:38:49 INF]  alphametics                (1.2.0 -> 1.3.0)
[23:38:49 INF]  anagram                    (1.3.0 -> 1.4.0)
[23:38:49 INF]  atbash-cipher              (1.1.0 -> 1.2.0)
[23:38:49 INF]  book-store                 (1.3.0 -> 1.4.0)
[23:38:49 INF]  bracket-push               (1.3.0 -> 1.4.0)
(...)
```

`-s all` will list all exercises, grouped by category.

**--help** will no longer throw an error,  and there's now some usage information:

```
USAGE:
Generate all exercises:
dotnet run
Generate the exercise named 'foobar':
dotnet run --exercise foobar
List outdated exercises:
dotnet run --status outdated
```

### CODE REVIEW

File: generators/Exercise.fs [hunk 1]
```diff
@@ -57,6 +61,19 @@ type GeneratorExercise() =
         |> this.MapCanonicalData
         |> this.Render  
         |> this.WriteToFile
+        
+    member this.ReadVersion () =
+        (*
+            Read from the top of the file e.g.
+            "// This file was auto-generated based on version 1.2.0 of the canonical data."
+        *)
+        let isNum c = c >= '0' && c <= '9'
```
Comment (ErikSchierboom) [inline generators/Exercise.fs]: I think you could replace this with `System.Char.IsDigit`

File: generators/Options.fs [hunk 2]
```diff
@@ -49,16 +68,22 @@ let private normalizeStatus status =
...
+    | None                 -> None
 
 let private mapOptions (options: CommandLineOptions) =
     { Exercise = normalizeExercise options.Exercise
       Status = normalizeStatus options.Status
       CanonicalDataDirectory = normalizeCanonicalDataDirectory options.CanonicalDataDirectory
       CacheCanonicalData = options.CacheCanonicalData }
 
+let conflictingStatusAndExerciseParams (parsed:Parsed<CommandLineOptions>) =
+    let s = parsed.Value.Status |> String.IsNullOrWhiteSpace |> not
```
Comment (ErikSchierboom) [inline generators/Options.fs]: Could you give this a more descriptive name? E.g. `statusParameterProvided` or `statusParameterSpecified`?

File: generators/Options.fs [hunk 3]
```diff
@@ -49,16 +68,22 @@ let private normalizeStatus status =
...
+    let e = parsed.Value.Exercise |> String.IsNullOrWhiteSpace |> not
```
Comment (ErikSchierboom) [inline generators/Options.fs]: Could you give this a more descriptive name? E.g. `exerciseParameterProvided` or `exerciseParameterSpecified`?

File: generators/Options.fs [hunk 4]
```diff
@@ -3,23 +3,42 @@ module Generators.Options
...
+          static member Defaults () =
+              {
+                Exercise = null
+                Status = ""
+                CanonicalDataDirectory = null
+                CacheCanonicalData = false
+              }
+              
+          [<Usage(ApplicationAlias = "dotnet run")>]
+          static member Examples =
```
Comment (ErikSchierboom) [inline generators/Options.fs]: This is great!

File: generators/Program.fs [hunk 5]
```diff
@@ -48,21 +36,32 @@ let private regenerateTestClasses options =
...
+            
+                      
 
 [<EntryPoint>]
 let main argv = 
     Logging.setupLogger()
 
     match parseOptions argv with
+    | Ok(options) when options.Status.IsSome && options.Exercise.IsSome -> 
+        Log.Error("Can't have both -s and -e.")
```
Comment (ErikSchierboom) [inline generators/Program.fs]: Maybe reword this slightly, as the user could also have specified the arguments as `--exercise` and `--status`?

File: generators/Program.fs [hunk 6]
```diff
@@ -48,21 +36,32 @@ let private regenerateTestClasses options =
...
+        1
+    | Ok(options) when options.Status.IsSome -> 
+        Reporting.listExercises options
```
Comment (ErikSchierboom) [inline generators/Program.fs]: Maybe open the `Reporting` module, so this can be become `listExercises`?

File: generators/Reporting.fs [hunk 7]
```diff
@@ -0,0 +1,92 @@
+module Generators.Reporting
+
+open Serilog
+open CanonicalData
+open Exercise
+open Options
+       
+let private isOutdated (exercise: GeneratorExercise) options (parseCanonicalData':string -> CanonicalData) =
+    let cData = parseCanonicalData' exercise.Name
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you perhaps come up with a more descriptive identifier?

File: generators/Reporting.fs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    match cData.Version,exercise.ReadVersion() with
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you add a space after the comma?

File: generators/Reporting.fs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    | canonVersion,exerciseVersion when canonVersion.Equals exerciseVersion -> false
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you add a space after the comma?

Comment (ErikSchierboom) [inline generators/Reporting.fs]: I'm not sure if people would understand `canonVersion`, maybe `canonicalDataVersion` would be better?

File: generators/Reporting.fs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    | _,_ -> true
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: I think you can replace this with just a single underscore, there is no need to pattern match on the tuple's contents.

File: generators/Reporting.fs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    | UpToDate of string
+    | Outdated of string * string * string
+    
+let nameFromSummaryType = function
+    | Custom n -> n
+    | Unimplemented  n -> n
+    | MissingData n -> n
+    | Deprecated n -> n
+    | UpToDate n -> n
+    | Outdated (n,_,_) -> n
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you add a spaces after the commas?

File: generators/Reporting.fs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    
+let private summarizeExercise options (parseCanonicalData':string -> CanonicalData) (exercise:Exercise)  =
+
+    match exercise with
+    | Exercise.Custom custom ->                 SummaryTypes.Custom custom.Name
+    | Exercise.Unimplemented unimplemented ->   SummaryTypes.Unimplemented unimplemented.Name
+    | Exercise.MissingData missingData ->       SummaryTypes.MissingData missingData.Name
+    | Exercise.Deprecated deprecated ->         SummaryTypes.Deprecated deprecated.Name
+    | Exercise.Generator generator ->
+        let cData = parseCanonicalData' generator.Name
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you perhaps come up with a more descriptive identifier?

File: generators/Reporting.fs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        
+        match generator.ReadVersion (), cData.Version with
+        | a,b when a.Equals b ->  UpToDate generator.Name
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you add a spaces after the commas? Also there is one space too many after the `->`

File: generators/Reporting.fs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        | a,b -> Outdated (generator.Name,a,b)
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you add a space after the commas?

File: generators/Reporting.fs (hunk 13 above)
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you perhaps come up with more descriptive identifiers for `a` and `b`?

File: generators/Reporting.fs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+                          | SummaryTypes.MissingData _ -> "are missing canonical data"
+                          | SummaryTypes.Deprecated _ -> "are deprecated"
+                          | SummaryTypes.UpToDate _ -> "are up to date"
+                          | SummaryTypes.Outdated _ -> "are outdated"
+                          
+        Log.Information("{num} exercises {description}:", group.Length, description)
+        
+        group |> List.iter (fun exercise -> 
+            match exercise with
+            | SummaryTypes.Outdated (name,oldVer,newVer) -> Log.Information(" {name}{indent}({oldVer} -> {newVer})", name, indentAfter name, oldVer, newVer)
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you remove the space before the exercise name in the logging statement?

File: generators/Reporting.fs [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+            | _ -> Log.Information(" {name}", nameFromSummaryType exercise)
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you remove the space before the exercise name in the logging statement?

File: generators/Reporting.fs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        )
+            
+        printfn ""
+
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you remove this unneeded empty line?

File: generators/Reporting.fs [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+                          
+        Log.Information("{num} exercises {description}:", group.Length, description)
+        
+        group |> List.iter (fun exercise -> 
+            match exercise with
+            | SummaryTypes.Outdated (name,oldVer,newVer) -> Log.Information(" {name}{indent}({oldVer} -> {newVer})", name, indentAfter name, oldVer, newVer)
+            | _ -> Log.Information(" {name}", nameFromSummaryType exercise)
+        )
+            
+        printfn ""
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: What is the purpose of this empty line being output?

File: generators/Reporting.fs (hunk 15 above)
Comment (ErikSchierboom) [inline generators/Reporting.fs]: I was thinking that maybe `Log.Warning` would be a better logging level perhaps for outdated exercises? What do you think?

File: generators/Reporting.fs [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    
+    let exerciseGroups = allExercises
+                            |> List.filter (filterByStatus options parseCanonicalData')
+                            |> List.map (summarizeExercise options parseCanonicalData')
+                            |> List.groupBy (fun x -> x.GetType())
+    
+    exerciseGroups |> List.iter (fun (groupType,group) ->
+        let description = match group.Head with
+                          | SummaryTypes.Custom _ -> "have customized tests"
+                          | SummaryTypes.Unimplemented _ -> "are missing test generator"
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: `missing test generator` -> `missing a test generator`

File: generators/Reporting.fs [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+        
+let listExercises options =
+    Log.Information(sprintf "Listing exercises with status %s..." (options.Status.Value.ToString()))
+    
+    let parseCanonicalData' = CanonicalData.parseCanonicalData options
+    
+    let allExercises = createExercises options
+    
+    let longestNameLength = allExercises |> List.map exerciseName |> List.map (fun e -> e.Length) |> List.max
+    let indentAfter (name:string) = String.replicate (longestNameLength+2-name.Length) " "
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you add spaces around the operators (+, -)?

Comment (ErikSchierboom) [inline generators/Reporting.fs]: I think you could simplify this to `"".PadRight(longestNameLength+2-name.Length)`, but please use whichever version you like best.

File: generators/Reporting.fs [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+        | a,b -> Outdated (generator.Name,a,b)
+        
+let listExercises options =
+    Log.Information(sprintf "Listing exercises with status %s..." (options.Status.Value.ToString()))
+    
+    let parseCanonicalData' = CanonicalData.parseCanonicalData options
+    
+    let allExercises = createExercises options
+    
+    let longestNameLength = allExercises |> List.map exerciseName |> List.map (fun e -> e.Length) |> List.max
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: This line is a bit long, maybe you could reformat to have the pipes on separate lines? Like this:

```fsharp
let longestNameLength = 
    allExercises
    |> List.map exerciseName
    |> List.map (fun e -> e.Length)
    |> List.max
```

File: generators/Reporting.fs [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+    let indentAfter (name:string) = String.replicate (longestNameLength+2-name.Length) " "
+    
+    let exerciseGroups = allExercises
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you slightly reformat this to have the pipes start at the same level as the expression that is being piped into it?

```fsharp
let exerciseGroups = 
    allExercises
    |> List.filter (filterByStatus options parseCanonicalData')
    |> List.map (summarizeExercise options parseCanonicalData')
    |> List.groupBy (fun x -> x.GetType())
```

File: generators/Reporting.fs [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+                            |> List.filter (filterByStatus options parseCanonicalData')
+                            |> List.map (summarizeExercise options parseCanonicalData')
+                            |> List.groupBy (fun x -> x.GetType())
+    
+    exerciseGroups |> List.iter (fun (groupType,group) ->
+        let description = match group.Head with
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you slightly reformat this to have the pipes start at the same level as the expression that is being piped into it?

```fsharp
let description =
    match group.Head with
    | SummaryTypes.Custom _ -> "have customized tests"
    | SummaryTypes.Unimplemented _ -> "are missing test generator"
    | SummaryTypes.MissingData _ -> "are missing canonical data"
    | SummaryTypes.Deprecated _ -> "are deprecated"
    | SummaryTypes.UpToDate _ -> "are up to date"
    | SummaryTypes.Outdated _ -> "are outdated"
```

File: generators/Reporting.fs [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+                          | SummaryTypes.Custom _ -> "have customized tests"
+                          | SummaryTypes.Unimplemented _ -> "are missing test generator"
+                          | SummaryTypes.MissingData _ -> "are missing canonical data"
+                          | SummaryTypes.Deprecated _ -> "are deprecated"
+                          | SummaryTypes.UpToDate _ -> "are up to date"
+                          | SummaryTypes.Outdated _ -> "are outdated"
+                          
+        Log.Information("{num} exercises {description}:", group.Length, description)
+        
+        group |> List.iter (fun exercise ->
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Maybe extract the listing of a single exercise to a helper function?

File: generators/Reporting.fs (hunk 15 above)
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Could you add spaces in the tuple expressions?

File: generators/Reporting.fs [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+    
+    let longestNameLength = allExercises |> List.map exerciseName |> List.map (fun e -> e.Length) |> List.max
+    let indentAfter (name:string) = String.replicate (longestNameLength+2-name.Length) " "
+    
+    let exerciseGroups = allExercises
+                            |> List.filter (filterByStatus options parseCanonicalData')
+                            |> List.map (summarizeExercise options parseCanonicalData')
+                            |> List.groupBy (fun x -> x.GetType())
+    
+    exerciseGroups |> List.iter (fun (groupType,group) ->
```
Comment (ErikSchierboom) [inline generators/Reporting.fs]: Maybe extract the listing of an exercise group to a helper function?

File: generators/Options.fs (hunk 4 above)
Comment (jrr) [reply generators/Options.fs]: This is my favorite feature of the CommandLine library. It's so slick to specify the usage programmatically, and get the string generated for you. Much less likely to rot the way manually-maintained documentation does.

File: generators/Reporting.fs (hunk 15 above)
Comment (jrr) [reply generators/Reporting.fs]: If you prefer, but I meant it to be a small amount of indentation to visually distinguish the exercise list under the category summary. For now I added a couple more spaces. Does that look better?
```
[10:45:16 INF] 20 exercises are outdated:
[10:45:16 INF]    allergies                  (1.1.0 -> 1.2.0)
[10:45:16 INF]    alphametics                (1.2.0 -> 1.3.0)
[10:45:16 INF]    anagram                    (1.3.0 -> 1.4.0)
[10:45:16 INF]    atbash-cipher              (1.1.0 -> 1.2.0)
[10:45:16 INF]    book-store                 (1.3.0 -> 1.4.0)
[10:45:16 INF]    bracket-push               (1.3.0 -> 1.4.0)
(...)
```

File: generators/Reporting.fs (hunk 17 above)
Comment (jrr) [reply generators/Reporting.fs]: This white space was deliberate, as well, to separate the groups when you print them all:
```
[10:45:16 INF] 4 exercises are deprecated:
[10:45:16 INF]    binary
[10:45:16 INF]    hexadecimal
[10:45:16 INF]    octal
[10:45:16 INF]    trinary

[10:45:16 INF] 1 exercises have customized tests:
[10:45:16 INF]    diamond

[10:45:16 INF] 3 exercises are missing test generator:
[10:45:16 INF]    grade-school
[10:45:16 INF]    pythagorean-triplet
[10:45:16 INF]    sgf-parsing
```

(but if you prefer, I'll remove it)

File: generators/Reporting.fs (hunk 11 above)
Comment (jrr) [reply generators/Reporting.fs]: The community really needs some better lint tooling and autoformatting for this kind of stuff 😐

File: generators/Reporting.fs (hunk 15 above)
Comment (jrr) [reply generators/Reporting.fs]: I'm not in the habit of using logging libraries for small command-line tools like this. What I think is that we could remove the first column of output entirely ;)

If this were a warning, then should some of the other categories be warnings too?

File: generators/Reporting.fs (hunk 11 above)
Comment (ErikSchierboom) [reply generators/Reporting.fs]: Absolutely agree! Maybe we should take a look at integrating [fantomas](https://github.com/fsprojects/fantomas) sometime.

File: generators/Exercise.fs (hunk 1 above)
Comment (jrr) [reply generators/Exercise.fs]: Ah! I knew there had to be a better way.

File: generators/Reporting.fs (hunk 15 above)
Comment (jrr) [reply generators/Reporting.fs]: Once I started using `Log.Information("{foo} {bar}", foo, bar)` instead of `Log.Information(sprintf "%s %s" foo bar)`, I came to really like the colors, too.

Maybe we could configure Serilog to leave the first column off, but keep the benefit of colorizing the output?

File: generators/Program.fs (hunk 5 above)
Comment (ErikSchierboom) [reply generators/Program.fs]: Sounds good!

File: generators/Reporting.fs (hunk 15 above)
Comment (ErikSchierboom) [reply generators/Reporting.fs]: Sounds like a plan! If you wish to include that in this PR, that would be fine. A separate PR would be better :)

Comment (jrr) [reply generators/Reporting.fs]: I just pushed up the single-space version, applied to both `-s` and `-e`. It'd be easy to change later if we want.

**Comment by ErikSchierboom** [comment]
Merged! This is a fantastically useful PR, thanks again!

File: generators/Reporting.fs (hunk 15 above)
Comment (robkeim) [reply generators/Reporting.fs]: For spacing things like this I don't have a strong preference one way over another, but I do have a strong preference for consistency :)
